keps/sig-node/3675-resource-plugin-manager/README.md

```diff
@@ -0,0 +1,1328 @@
+  and provide an interface while not having vendor-specific code within Kubelet.
+* Differentiate between types of cores and memory. <br>
+  Note - Dynamic resouce allocation does this with regular resources today.  We seek to
+  extend this ability.
+* Differentiate between different configurations of cores and memory, for instance cores
+  designated as performance versus those designated as efficiency cores 
+* Have custom plugins to optimize for particular types of workloads.  These plugins
+  may be built for performance, power reduction, cpu savings, et cetera.  <br>
+  Note:  Currently, there are very limited sets of available topology policies.  Every
+  new policy must be approved and be lockstep with the Kuberenetes release process.
+* Be able to hot-plug and test new resource managers.  
+* Be able to remove some of the complexity with current setup and, over time, reduce
+  the amount of code contained within Kubelet.  Instead, build a library with specific
+  needs.
+* Have a faster path to desired changes, without potentially impacting the core of 
+  Kubernetes with every policy change. <br>
+  Note that current solutions have been cropping up to allow for resource management
+  outside the core Kubelet. These solutions require turning off the Kubelet functionality 
+  and overriding current Kubelet allocation. We should provide a path otherwise. Many of these solutions are custom to particular companies. Many are not open source. The community should give a process to allow this functionality in core kubelet at a granular level, same as we have in many systems, such as HPC or telco.
+* Leverage additional parts from the pod spec information which could help make allocation decisions
+without the need to query the k8s controlplane.
+* Be able to do research, with minimum toil, on new policies and resource management strategies
+
+This design will also use gRPC, which is used for many other pluggable components within Kubernetes.  
+The hope is that it will also, as a side effect, allow a path to simplify the Kubelet as it is 
+today and move the complexity into external plugins.
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+#### Alpha Goals
+
+1.	Provide Initial infrastructure to support CCI Driver Plugins and k8s Deployments requiring CCI Drivers after scheduling
+2.	Support seamless k8s start ( no plugins required) through the intermediate CCI CPU management policy (functionality will be moved in beta to None and Static policies)
+3.	Support standard pod deployments not requiring CCI Drivers through CPU Manager: best-effort QoS(stretch goal will include static handling of burstable and guaranteed QoS)
+4.	CCI Driver results passed to CPU Manager (i.e Resource sets) have to support cpusets and memory affinity, .. 
+5.	Consistent on-none state handling between CCI Driver and CPU Manager (race-free)
```

Comment:
This section will need some tweaks based on the approach we discussed above. The state management would need to happen between the legacy pods that want to continue using `none` policy and the new ones (with CCIDriver indicated in the pod spec) that want to be allocated CPUs via the CCI driver.
How is the default CPU pool impacted by the presence of the pods that are allocated by CCI drivers? Does it shrink when new pods are allocated CPUs via the CCIDriver?

Reply:
> The state management would need to happen between the legacy pods that want to continue using none policy and the new ones (with CCIDriver indicated in the pod spec) that want to be allocated CPUs via the CCI driver.

I think *all* pods should either go through the existing CPUManager (when the CCI feature gate is disabled) or down a completely different code path for CCI implemented CPUManagers (when the CCI feature gate is enabled).

In that way there is no "opting in" to be managed by a CCI implemented CPUManager on a pod-by-pod basis -- *all* pods will be managed by the CCI implemented CPUManager if the CCI feature gate is enabled and *all* pods will be managed by the traditional CPUManager if it is disabled.

The whole notion of "erroring out" if `none` is not set for the traditional CPUManager is just to ensure that the `kubelet` fails fast if the CCI feature gate is enabled and a CPUManager policy other than `none` is set. It doesn't mean that some pods go through the traditional CPUManager and some go through a CCI implemented CPUManager -- it should be all or none.

The one caveat here being that there is a bootstrapping / failure case scenario to consider when the pod for a CCI implemented CPUManager plugin is not (yet) running or has gone offline for some reason. For alpha I would argue we should just log this in the kubelet, start the pod anyway, and continue.

We can consider what the "right" thing to do here is at a later stage once we have a better grasp on running these plugins in general.

Reply:
Added for beta, for failure.